Source/engine/random.hpp

```diff
@@ -57,4 +58,18 @@ int32_t AdvanceRndSeed();
  */
 int32_t GenerateRnd(int32_t v);
 
+/**
+ * @brief Picks one of the elements in the list randomly.
+ *
+ * @param values The values to pick from
+ * @return A random value from the 'values' list.
+ */
+template <typename T>
+const T PickRandomlyAmong(const std::initializer_list<T> &values)
+{
+	auto index = GenerateRnd(values.size());
+
+	return *(values.begin() + index);
```

Comment:
What's the intent when `GenerateRnd()` returns a negative value? This code will attempt to read past the head of the list. Need to document the exceptional case and either clamp to 0 or recognise callers that may rely on reading out of bounds to match vanilla logic.

Reply:
Wait... `GenerateRnd` can return a negative value? The hell....... I'll take a look.

Reply:
It's a rare case but that's what caused the segfault in the quest tests in #2751. See https://github.com/diasurgical/devilutionX/blob/946ae987a31c239499a59525c15df99a904ac842/Source/engine/random.hpp#L44-L58 and the comment on AdvanceRndSeed for more detail.

Reply:
If `GenerateRnd` can create a negative value, the original code would generate a `ItemType::Misc` item on a weapon rack.... is that really a thing? I've never heard of someone getting that before.

Reply:
Anyways....... clamped the value to 0. Let me know if that's fine.

Reply:
Yep, that looks good. For weapon racks the old code (and the change you made) was safe because `GenerateRnd(4)` will always return 0, 1, 2, or 3.

If for example you tried to add shields to the potential drops then `GenerateRnd(5)` can return 0, 1, 2, 3, 4, or -2. The -2 value would correspond to ItemType::None which isn't handled by the switch statement so the old code would call CreateTypeItem with an uninitialised `weaponType` variable. The first version of your PickRandomlyAmong code would try dereference `(values.begin() - 2)` which may or may not crash.

Reply:
That's pretty crazy......... I guess at least it _can_ return all of the values in the range, so the clamping works, but biases the result against position 0 in that case?

Wouldn't it be possible to just try generating another number if the value returns negative as an implementation detail on `GenerateRnd`? Or would advancing the seed in the process mess up things?

Reply:
Clamping to zero also has the benefit of correcting a slight bias that would otherwise be introduced. Definitely do not want to reroll as that would break things that rely on a specific number of RNG calls like item generation.

Reply:
Wouldn't clamping it make it biased to 0?